### PR TITLE
Insert a blank line between consecutive compiler messages

### PR DESCRIPTION
**Pull request by gasche**
This PR is inspired by user feedback from @mimoo on compiler error messages which originated in #11539.

Currently OCaml tools will include several messages concatenated together -- several warnings, possibly followed by one error -- and it may not be obvious especially for new users to visually split the output in several reports. 

This PR introduces a blank line between consecutive reports.

(cc @Octachron)

### Batch compiler example

Before:

```
File "b_bad.ml", lines 13-14, characters 29-28:
13 | .............................function
14 |     A.X s -> print_endline s
Error (warning 8 [partial-match]): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Y
File "b_bad.ml", line 18, characters 11-14:
18 | let () = f A.y
                ^^^
Error: Unbound value A.y
```

After:

```
File "b_bad.ml", lines 13-14, characters 29-28:
13 | .............................function
14 |     A.X s -> print_endline s
Error (warning 8 [partial-match]): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Y

File "b_bad.ml", line 18, characters 11-14:
18 | let () = f A.y
                ^^^
Error: Unbound value A.y
```

### Toplevel example

Before:

```
# type _ t = Int : int -> int t | Bool : bool -> bool t;;
type _ t = Int : int -> int t | Bool : bool -> bool t
# let test : int t -> int = function Int n -> n | _ -> 0;;
Warning 4 [fragile-match]: this pattern-matching is fragile.
It will remain exhaustive when constructors are added to type t.
Warning 56 [unreachable-case]: this match case is unreachable.
Consider replacing it with a refutation case '<pat> -> .'
val test : int t -> int = <fun>
```

After:

```
# type _ t = Int : int -> int t | Bool : bool -> bool t;;
type _ t = Int : int -> int t | Bool : bool -> bool t
# let test : int t -> int = function Int n -> n | _ -> 0;;
Warning 4 [fragile-match]: this pattern-matching is fragile.
It will remain exhaustive when constructors are added to type t.

Warning 56 [unreachable-case]: this match case is unreachable.
Consider replacing it with a refutation case '<pat> -> .'

val test : int t -> int = <fun>
```

### Implementation

Most of the logic is hidden in `Location`, but unfortunately there are other places in the compiler that print user messages than `Location.print_report`, so we need to break the abstraction a bit and expose a function "I'm about to print a new message" to be called by other modules. Not super nice, but not too bad either -- and very short.

### Note

In the original report, the issue was at the build-system level (the build system would concatenate several errors coming from different files); I reported this issue on dune in https://github.com/ocaml/dune/issues/6158 , and this is getting fixed by @esope in https://github.com/ocaml/dune/pull/6823. The present PR is about messages concatenated by the compiler itself.

### CODE REVIEW

File: parsing/location.ml
Comment (mimoo) [inline parsing/location.ml]: "usere input"

Comment (gasche) [reply parsing/location.ml]: Thanks, fixed.

**Comment by Octachron** [comment]
The code change looks fine to me: it seems fine to have a function that is used to mark the start of separate user messages. I expect that if we move to a structured log for the compiler functions those function calls will be naturally subsumed by the more general API.

**Comment by gasche** [comment]
This reads like an approving review, without the approval stamp. What's your thought on what this PR needs next?

**Comment by Octachron** [comment]
I am reading the testsuite changes before giving my approval.

File: testsuite/tests/warnings/deprecated_module_use.compilers.reference
Comment (Octachron) [inline testsuite/tests/warnings/deprecated_module_use.compilers.reference:9]: The second blank line here comes for the deprecation message, which suggest that this PR might slightly affect the design of deprecation message.

File: testsuite/tests/warnings/deprecated_warning_specs.ml
Comment (Octachron) [inline testsuite/tests/warnings/deprecated_warning_specs.ml]: This newline is from an empty toplevel phrase.

Comment (gasche) [reply testsuite/tests/warnings/deprecated_warning_specs.ml]: Ah, I guess my move to a `signature option` should be changed back into "is the empty signature?" to account for other ways the signature could be empty. I will do this.

**Comment by Octachron** [comment]
The regexp of ocamltex's `catch_warning` needs to be updated to be able to ignore blank lines before the warning.

File: testsuite/tests/warnings/deprecated_warning_specs.ml
Comment (gasche) [reply testsuite/tests/warnings/deprecated_warning_specs.ml]: I fixed this (now the toplevel-specific commit has a much shorter diff) and rebased. Thanks for the eagle eyes!

**Comment by gasche** [comment]
> The regexp of ocamltex's `catch_warning` needs to be updated to be able to ignore blank lines before the warning.

Urgh, would you maybe be willing to do this part yourself? (You should have push rights on my fork.)

**Comment by Octachron** [comment]
The fix for ocamltex is available at Octachron@5b8852a16979af60 .

**Comment by gasche** [comment]
Thanks! I added it to the PR.

**Comment by gasche** [comment]
There were some bits of testsuite changes left to promote, I did this and kept the fixup commits around for now for easier reviewing.

**Review by Octachron — APPROVED** [review]
The testsuite is passing and adding some negative spaces to the compiler messages is a good idea.

**Pull request merged**
